lib/Core/BuildEngine.cpp

```diff
@@ -1398,6 +1402,14 @@ class BuildEngineImpl : public BuildDBDelegate {
   /// @{
 
   const ValueType& build(const KeyType& key) {
+    // Protect the engine against invalid concurrent use.
+    if (buildRunning.exchange(true)) {
+      delegate.error("build engine busy");
+      static ValueType emptyValue{};
+      return emptyValue;
+    }
+    scope_defer(buildRunning = false);
```

Comment:
This feels like a lot of potentially surprising "infrastructure" for a single use. Are you sure we will leverage this complexity?

Reply:
Also, would it not be possible to build an RAII based defer which takes a lambda? Does LLVM have something like that?

Reply:
I was hoping llvm would have something like it, but came up empty.  Maybe I just missed it? 

The ScopeDeferImpl is an RAII that takes a lambda (or other callable).  The macro just wraps it for easier use. 

I think this meme/util will be useful elsewhere, but it is not strictly necessary.  The number of exits is small and unlikely to grow in the method, so we could just litter it around.

Reply:
Maybe if I drop the `Impl` part of the name, it would be more clear that the class itself could be used directly and that the macro is just syntactic sugar.

Reply:
I’d probably do that, and drop the macro, to start with.

@DougGregor does Swift have something like this?

Reply:
I did rename it already in the current version.  I like the macro, but can defer* on it when back at my keyboard. 

* pun totally intended

Reply:
> does Swift have something like this?

Yeah, `@autoclosure`

Reply:
The Swift compiler's implementation is at https://github.com/apple/swift/blob/master/include/swift/Basic/Defer.h

Reply:
Well that looks quite familiar.  I like the use of braces in the macro form.

Reply:
I cede all my review feedback to follow Swift's precedent